### PR TITLE
agdarsec: init at 0.4.1

### PR DESCRIPTION
**Pull request by turion**
###### Motivation for this change

Add the https://github.com/gallais/agdarsec/ library to nixpkgs
u
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@alexarice

### CODE REVIEW

**Comment by turion** [comment]
I'm currently having trouble to build:

```
nix-repl> :b agdaPackages.agdarsec
builder for '/nix/store/bwfq8fswhs2l9a65zkh9yranhpv331q5-agdarsec-v0.3.0.drv' failed with exit code 1; last 10 log lines:
  building
  Checking index (/build/source/index.agda).
   Checking Text.Parser.Types (/build/source/src/Text/Parser/Types.agda).
   Checking Text.Parser.Combinators (/build/source/src/Text/Parser/Combinators.agda).
    Checking Induction.Nat.Strong (/build/source/src/Induction/Nat/Strong.agda).
    Checking Data.List.Any (/nix/store/vrr50yf58l2k36vw9clm43hfahn7msfh-standard-library-1.3/src/Data/List/Any.agda).
  Failed to write interface /nix/store/vrr50yf58l2k36vw9clm43hfahn7msfh-standard-library-1.3/src/Data/List/Any.agdai.
  /build/source/src/Text/Parser/Combinators.agda:17,8-28
  /nix/store/vrr50yf58l2k36vw9clm43hfahn7msfh-standard-library-1.3/src/Data/List/Any.agdai:
  openBinaryFile: permission denied (Permission denied)
[0 built (1 failed)]
error: build of '/nix/store/bwfq8fswhs2l9a65zkh9yranhpv331q5-agdarsec-v0.3.0.drv' failed
```

Strangely, it wants to recheck standard library files. Any idea why it might want to do that? Maybe `agdarsec` changes some build flag in a way that causes recompilation?

**Comment by alexarice** [comment]
I think this suggests that something in standard library is not being compiled

**Comment by turion** [comment]
@alexarice You're right!

```
$ ls /nix/store/vrr50yf58l2k36vw9clm43hfahn7msfh-standard-library-1.3/src/Data/List/Any.*
/nix/store/vrr50yf58l2k36vw9clm43hfahn7msfh-standard-library-1.3/src/Data/List/Any.agda
$ cat /nix/store/vrr50yf58l2k36vw9clm43hfahn7msfh-standard-library-1.3/Everything.agda | grep List.Any
-- Data.List.Any.Membership instantiated with propositional equality,
```

It seems the everything file is generated incorrectly.

**Comment by alexarice** [comment]
See https://github.com/agda/agda-stdlib/blob/master/src/Data/List/Any.agda.

The issue is that agdarsec relies on a deprecated module in standard library, and deprecated modules aren't built. Likely agdarsec needs updating to use the undeprecated module

**Comment by turion** [comment]
Hah. After monkey-patching `Data.List.Any` into the everything file, I get:

```
$ nix repl .
Welcome to Nix version 2.3.7. Type :? for help.

Loading '.'...
Added 12573 variables.

nix-repl> :b agdaPackages.agdarsec
builder for '/nix/store/qa9hmr88h2ib9zc5qf06kb909a1mqibd-standard-library-1.3.drv' failed with exit code 1; last 10 log lines:
    Checking Text.Tabular.Vec (/build/source/src/Text/Tabular/Vec.agda).
   Checking Text.Tree.Linear (/build/source/src/Text/Tree/Linear.agda).
   Checking Data.List.Any (/build/source/src/Data/List/Any.agda).
  /build/source/Everything.agda:1985,1-21
  Data.List.Any was deprecated in v1.0.
  Use Data.List.Relation.Unary.Any instead.
  when scope checking the declaration
    import Data.List.Any
```

So it's actually a deprecation warning.

**Comment by turion** [comment]
@alexarice Yes, right. This is fixed on `master` I think.

**Comment by AndersonTorres** [comment]
@turion do you think it is a good idea to use master, then?

**Comment by turion** [comment]
@AndersonTorres It would be ok in principle until the next release, but unfortunately `master` doesn't work out of the box either because of version mismatches. I'm working on a workaround.

File: pkgs/development/libraries/agda/agdarsec/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  buildInputs = [ standard-library ];
+
+  meta = with stdenv.lib; {
+    inherit (src.meta) homepage;
+    description = "Total Parser Combinators in Agda";
+    license = stdenv.lib.licenses.gpl3;
+    platforms = stdenv.lib.platforms.unix;
+    maintainers = with maintainers; [ turion ];
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/agda/agdarsec/default.nix:25]:
```suggestion
  meta = with stdenv.lib; {
    inherit (src.meta) homepage;
    description = "Total Parser Combinators in Agda";
    license = licenses.gpl3;
    platforms = platforms.unix;
    maintainers = with maintainers; [ turion ];
  };
```

File: pkgs/applications/editors/vscode/vscodium.nix [hunk 2]
```diff
@@ -11,8 +11,8 @@ let
   archive_fmt = if system == "x86_64-darwin" then "zip" else "tar.gz";
 
   sha256 = {
-    x86_64-linux = "19y25yfkls53w4qlmipfvjig7zykgwx1010ny58k339fv181vdyq";
-    x86_64-darwin = "1ak3pfvwdg51hcv2kyqpnhzkl7k23k5qk197sf6rv02kgwan7pxl";
+    x86_64-linux = "0fhwqif9021sg22n8qqn3m2fml7sjb03ydgbakg1021i3y8zl599";
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/vscodium.nix:14]: Please remove this from this PR.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/editors/vscode/vscodium.nix (hunk 2 above)
Comment (turion) [reply pkgs/applications/editors/vscode/vscodium.nix:14]: Thanks for spotting.

**Comment by turion** [comment]
I'm currently stuck on this error:

```
nix-repl> :b agdaPackages.agdarsec
builder for '/nix/store/djhsxg2av3g3wdvbcimq29dj0j602hkf-agdarsec-v0.3.0.drv' failed with exit code 1; last 10 log lines:
    /nix/store/1b7jp08xlcbvankbf8aml162zmid00la-standard-library-1.4/src/Expr.agda
    /nix/store/1b7jp08xlcbvankbf8aml162zmid00la-standard-library-1.4/src/Expr.lagda
    /build/source/src/Expr.agda
    /build/source/src/Expr.lagda
    /build/source/Expr.agda
    /build/source/Expr.lagda
    /nix/store/vkm07q5a493h356zmlqzvn7sxc4r0wwv-Agda-2.6.1.2-data/share/ghc-8.10.2/x86_64-linux-ghc-8.10.2/Agda-2.6.1.2/lib/prim/Expr.agda
    /nix/store/vkm07q5a493h356zmlqzvn7sxc4r0wwv-Agda-2.6.1.2-data/share/ghc-8.10.2/x86_64-linux-ghc-8.10.2/Agda-2.6.1.2/lib/prim/Expr.lagda
  when scope checking the declaration
    import Expr
[1 built (1 failed), 0.0 MiB DL]
error: build of '/nix/store/djhsxg2av3g3wdvbcimq29dj0j602hkf-agdarsec-v0.3.0.drv' failed
```

See https://github.com/gallais/agdarsec/issues/21.

**Comment by turion** [comment]
@alexarice @neosimsim do you want to have a look? I also extended the build support a little. Now you can add the `-i` option (include paths) through nix.

**Comment by turion** [comment]
@ofborg build agdaPackages.agdarsec
@ofborg test agdaPackages.standard-library.passthru.tests

File: pkgs/development/libraries/agda/agdarsec/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{ lib, mkDerivation, fetchFromGitHub
+, standard-library }:
+
+mkDerivation rec {
+  version = "0.4.1";
+  pname = "agdarsec";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/agda/agdarsec/default.nix:6]:
```suggestion
  pname = "agdarsec";
  version = "0.4.1";
```

Comment (turion) [reply pkgs/development/libraries/agda/agdarsec/default.nix:6]: Done.

**Comment by turion** [comment]
@ofborg build agdaPackages.agdarsec
@ofborg test agdaPackages.standard-library.passthru.tests

**Comment by turion** [comment]
@ofborg build agdaPackages.agdarsec
@ofborg test agdaPackages.standard-library.passthru.tests

**Comment by turion** [comment]
@alexarice @SuperSandro2000 fine like this?

**Comment by alexarice** [comment]
Looks fine to me. I'm surprised we haven't hit the include path problem before. Maybe we should do things like include the stdlib readmes? (this an be a separate pr)

File: pkgs/development/libraries/agda/agdarsec/default.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  };
+
+  everythingFile = "./index.agda";
+
+  includePaths = [ "src" "examples" ];
+
+  buildInputs = [ standard-library ];
+
+  meta = with lib; {
+    inherit (src.meta) homepage;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/agda/agdarsec/default.nix:22]:
```suggestion
    homepage = "https://gallais.github.io/agdarsec/";
```

**Review by SuperSandro2000 — COMMENTED** [review]
I have no clue about agda but I did not see anything which is worth a comment.

**Pull request merged**